crates/red_knot_python_semantic/src/types.rs

```diff
@@ -306,3 +312,359 @@ pub struct IntersectionType<'db> {
     // the intersection type does not include any value in any of these types
     negative: FxOrderSet<Type<'db>>,
 }
+
+#[allow(unused)]
+#[derive(Clone)]
+struct IntersectionTypeBuilder<'db> {
+    intersections: Vec<InnerIntersectionTypeBuilder<'db>>,
+    db: &'db dyn Db,
+}
+
+impl<'db> IntersectionTypeBuilder<'db> {
+    #[allow(dead_code)]
+    fn new(db: &'db dyn Db) -> Self {
+        Self {
+            db,
+            intersections: vec![InnerIntersectionTypeBuilder::new()],
+        }
+    }
+
+    fn empty(db: &'db dyn Db) -> Self {
+        Self {
+            db,
+            intersections: vec![],
+        }
+    }
+
+    #[allow(dead_code)]
+    fn add_positive(mut self, ty: Type<'db>) -> Self {
+        if let Type::Union(union) = ty {
+            union
+                .elements(self.db)
+                .iter()
+                .map(|elem| self.clone().add_positive(*elem))
+                .fold(
+                    IntersectionTypeBuilder::empty(self.db),
+                    |mut builder, sub| {
+                        builder.intersections.extend(sub.intersections);
+                        builder
+                    },
+                )
+        } else {
+            for inner in &mut self.intersections {
+                inner.add_positive(self.db, ty);
+            }
+            self
+        }
+    }
+
+    #[allow(dead_code)]
+    fn add_negative(mut self, ty: Type<'db>) -> Self {
+        if let Type::Union(union) = ty {
+            union
+                .elements(self.db)
+                .iter()
+                .map(|elem| self.clone().add_negative(*elem))
+                .fold(
+                    IntersectionTypeBuilder::empty(self.db),
+                    |mut builder, sub| {
+                        builder.intersections.extend(sub.intersections);
+                        builder
+                    },
+                )
+        } else {
+            for inner in &mut self.intersections {
+                inner.add_negative(self.db, ty);
+            }
+            self
+        }
+    }
+
+    #[allow(dead_code)]
+    fn build(self) -> Type<'db> {
+        let mut builder = UnionTypeBuilder::new(self.db);
+        for inner in self.intersections {
+            builder = builder.add(inner.build(self.db));
+        }
+        builder.build()
+    }
+}
+
+#[allow(unused)]
+#[derive(Debug, Clone)]
+struct InnerIntersectionTypeBuilder<'db> {
+    positive: FxOrderSet<Type<'db>>,
+    negative: FxOrderSet<Type<'db>>,
+}
+
+impl<'db> InnerIntersectionTypeBuilder<'db> {
+    fn new() -> Self {
+        Self {
+            positive: FxOrderSet::default(),
+            negative: FxOrderSet::default(),
+        }
+    }
+
+    /// Adds a positive type to this intersection.
+    fn add_positive(&mut self, db: &'db dyn Db, ty: Type<'db>) {
+        match ty {
+            Type::Intersection(inter) => {
```

Comment:
I guess `inter` is short for `intersection` here, but at first glance I assumed it was a typo and you meant to write `inner` :P